crates/wasmparser/src/binary_reader.rs

```diff
@@ -1325,19 +1332,68 @@ impl<'a> BinaryReader<'a> {
         Ok(self.buffer[self.position])
     }
 
-    fn val_type_from_byte(byte: u8) -> Option<ValType> {
-        match byte {
-            0x7F => Some(ValType::I32),
-            0x7E => Some(ValType::I64),
-            0x7D => Some(ValType::F32),
-            0x7C => Some(ValType::F64),
-            0x7B => Some(ValType::V128),
-            0x70 => Some(ValType::FuncRef),
-            0x6F => Some(ValType::ExternRef),
-            _ => None,
+    fn read_heap_type(&mut self) -> Result<HeapType> {
+        match self.peek()? {
+            0x70 => {
+                self.position += 1;
+                Ok(HeapType::Func)
+            }
+            0x6F => {
+                self.position += 1;
+                Ok(HeapType::Extern)
+            }
+            x if 0x80 & x == 0 => {
+                self.position += 1;
+                Ok(HeapType::Index(x as u32))
+            }
```

Comment:
I was reading over this again and I don't think that this binary decoding is quite right. It's actually unfortunately subtle I think what needs to happen here, but I think what you can do is call `read_var_s33` and then validate that the result is in-bounds aftewards (e.g. not negative as the binary encoding specifies).

Reply:
I think this basically needs to be similar in structure to `read_block_type` if you're looking to pattern-match